pylutron/__init__.py

```diff
@@ -655,6 +694,496 @@ def handle_update(self, args):
+        OPERATING_MODE = 3
+        FAN_MODE = 4
+        ECO_MODE = 5
+        ECO_OFFSET = 6
+        SCHEDULE_STATUS = 7
+        SYSTEM_MODE = 11
+        CALL_STATUS = 14
+        TEMP_CURRENT_C = 15
+        TEMP_SETPOINTS_C = 16
+
+    def __init__(
+        self,
+        lutron,
+        name,
+        integration_id,
+        uuid,
+        temp_units,
+        avail_op_modes,
+        avail_fan_modes,
+        avail_misc_features,
+        min_temp_cool,
+        max_temp_cool,
+        min_temp_heat,
+        max_temp_heat,
+        heat_cool_delta,
+    ):
+        """Initializes the HVAC controller."""
+        super(HVAC, self).__init__(lutron, name, uuid)
+        self._query_waiters = _RequestHelper()
+        self._integration_id = integration_id
+        self._lutron.register_id(HVAC._CMD_TYPE, self)
+        self._temp_units = "F" if temp_units == 1 else "C"
+        self._heat_cool_delta = heat_cool_delta if heat_cool_delta else 3
+        self._operating_modes = [slugify(mode, separator='_').upper() for mode in re.split(r"\s*,\s*", avail_op_modes)]
+        self._fan_modes = [slugify(mode, separator='_').upper() for mode in re.split(r"\s*,\s*", avail_fan_modes)]
+        self._avail_misc_features = [slugify(mode, separator='_').upper() for mode in re.split(r"\s*,\s*", avail_misc_features)]
+        self._eco_mode = False
+        self._min_temp_cool = min_temp_cool
+        self._max_temp_cool = max_temp_cool
+        self._min_temp_heat = min_temp_heat
+        self._max_temp_heat = max_temp_heat
+        self._call_status = HVAC.CallStatus.OFF.name
+        self._schedule_status = HVAC.ScheduleStatus.SCHEDULE_UNAVAILABLE.name
+        self._current_fan = HVAC.FanModes.NO_FAN.name
+        self._current_mode = HVAC.OperatingModes.OFF.name
+        self._current_temp_c = 0.0
+        self._current_temp_f = 0
+        self._setpoint_cool_c = 0.0
+        self._setpoint_heat_c = 0.0
+        self._setpoint_cool_f = 0
+        self._setpoint_heat_f = 0
+
+    def __str__(self):
+        """Returns a pretty-printed string for this object."""
+        return 'HVAC name: "%s" id: %d' % (self._name, self._integration_id)
+
+    def __repr__(self):
+        """Returns a stringified representation of this object."""
+        return str({'name': self._name, 'id': self._integration_id})
+
+    @property
+    def id(self):
+        """The integration id"""
+        return self._integration_id
+
+    @property
+    def avail_misc_features(self):
+        """The available miscellaneous features"""
+        return self._avail_misc_features
+    
+    @property
+    def heat_cool_delta(self):
+        """The required delta between heat and cool setpoints"""
+        return self._heat_cool_delta
+
+    def handle_update(self, args):
+        """Handles an event update for this object, e.g. temp level change."""
+        _LOGGER.debug("HVAC handle_update %d -- %s" % (self._integration_id, args))
+
+        def _u_current_temp_f(temp):
```

Comment:
Something about this makes me uneasy enough that I'm contemplating alternative arrangements.  I just don't have an excellent idea yet.

The one thing rattling around my head is what should the object model of an HVAC be?

Is it one object with a bunch of properties?  Or is the fan an object, and the operating mode an object, etc... 

It feels like there's a lot of repeated boilerplate below that could be factored out and make it easier to read and parse.  Like an HVACTemperatire class that can deal with the current temperature and setpoints (you end up with one instance for Farenheight and one for Celcius),  and an HVACModes class, etc...

Reply:
I'm not opposed to the idea at all but that might also introduce more complexity and not save much actual code in the long run.  Additionally, when consuming this in HA there are crossovers between modes, fans, and presets.  Having multiple objects makes it more complicated to consume and increases the code size there.